File: backend/src/main/java/com/interview/service/PdfReportService.java
package com.interview.service;

import com.interview.model.Resume;
import lombok.extern.slf4j.Slf4j;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.pdmodel.PDPageContentStream;
import org.apache.pdfbox.pdmodel.common.PDRectangle;
import org.apache.pdfbox.pdmodel.font.PDType1Font;
import org.apache.pdfbox.pdmodel.font.Standard14Fonts;
import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Service for generating styled PDF reports using PDFBox 3.x
 */
@Slf4j
@Service
public class PdfReportService {

    // Helper context class to carry document state across page breaks
    private static class DocumentContext {
        final PDDocument document;
        PDPage currentPage;
        PDPageContentStream currentStream;
        float currentY;
        final float margin = 50;
        final float pageWidth;
        final float pageHeight;
        final float printableWidth;
        
        DocumentContext(PDDocument document) throws IOException {
            this.document = document;
            this.currentPage = new PDPage(PDRectangle.A4);
            this.document.addPage(this.currentPage);
            
            this.pageWidth = currentPage.getMediaBox().getWidth();
            this.pageHeight = currentPage.getMediaBox().getHeight();
            this.printableWidth = pageWidth - (2 * margin);
            this.currentY = pageHeight - margin;
            
            this.currentStream = new PDPageContentStream(document, currentPage);
        }
        
        void checkNewPage(float requiredHeight) throws IOException {
            // If the content would exceed bottom margin (50), start a new page
            if (currentY - requiredHeight < 50) {
                currentStream.close();
                
                currentPage = new PDPage(PDRectangle.A4);
                document.addPage(currentPage);
                
                currentStream = new PDPageContentStream(document, currentPage);
                currentY = pageHeight - margin;
                
                log.debug("Added a new page due to height requirement: {}", requiredHeight);
            }
        }
        
        void close() throws IOException {
            if (currentStream != null) {
                currentStream.close();
            }
        }

        void setNonStrokingColor(int r, int g, int b) throws IOException {
            currentStream.setNonStrokingColor(r / 255f, g / 255f, b / 255f);
        }

        void setStrokingColor(int r, int g, int b) throws IOException {
            currentStream.setStrokingColor(r / 255f, g / 255f, b / 255f);
        }
    }

    public byte[] generateResumeReport(Resume resume) throws IOException {
        log.info("Generating PDF report for resume: {}", resume.getId());
        
        try (PDDocument document = new PDDocument()) {
            DocumentContext context = new DocumentContext(document);
            
            PDType1Font titleFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA_BOLD);
            PDType1Font headerFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA_BOLD);
            PDType1Font bodyFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA);
            PDType1Font bodyBoldFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA_BOLD);
            
            // 1. Draw Title Header Block
            context.checkNewPage(80);
            context.setNonStrokingColor(30, 41, 59); // Slate 800
            context.currentStream.addRect(context.margin, context.currentY - 60, context.printableWidth, 60);
            context.currentStream.fill();
            
            context.currentStream.beginText();
            context.currentStream.setFont(titleFont, 18);
            context.setNonStrokingColor(255, 255, 255); // White
            context.currentStream.newLineAtOffset(context.margin + 20, context.currentY - 38);
            context.currentStream.showText("RESUME ATS ANALYSIS REPORT");
            context.currentStream.endText();
            context.currentY -= 80;
            
            // 2. Draw Candidate Details & ATS Score
            context.checkNewPage(120);
            
            // Background card for candidate details
            context.setNonStrokingColor(248, 250, 252); // Slate 50
            context.currentStream.addRect(context.margin, context.currentY - 90, context.printableWidth, 90);
            context.currentStream.fill();
            
            // Border line for card
            context.setNonStrokingColor(203, 213, 225); // Slate 300
            context.currentStream.addRect(context.margin, context.currentY - 90, 5, 90);
            context.currentStream.fill();
            
            // Draw details text
            context.currentStream.beginText();
            context.currentStream.setFont(bodyBoldFont, 12);
            context.setNonStrokingColor(51, 65, 85); // Slate 700
            context.currentStream.newLineAtOffset(context.margin + 20, context.currentY - 25);
            context.currentStream.showText("Candidate Details");
            
            context.currentStream.setFont(bodyFont, 10);
            context.setNonStrokingColor(71, 85, 105);
            
            String name = resume.getParsedData() != null ? resume.getParsedData().getName() : "Not Found";
            String email = resume.getParsedData() != null ? resume.getParsedData().getEmail() : "Not Found";
            String phone = resume.getParsedData() != null ? resume.getParsedData().getPhone() : "Not Found";
            
            context.currentStream.newLineAtOffset(0, -18);
            context.currentStream.showText("Name: " + name);
            context.currentStream.newLineAtOffset(0, -15);
            context.currentStream.showText("Email: " + email);
            context.currentStream.newLineAtOffset(0, -15);
            context.currentStream.showText("Phone: " + phone);
            context.currentStream.endText();
            
            // ATS Score Badge (Inside candidate card box on the right)
            int score = (resume.getAnalysis() != null && resume.getAnalysis().getAtsScore() != null) 
                    ? resume.getAnalysis().getAtsScore() : 70;
            
            float scoreCardX = context.margin + 300;
            float scoreCardY = context.currentY - 15;
            
            // Score circle background container
            context.setNonStrokingColor(241, 245, 249); // Slate 100
            context.currentStream.addRect(scoreCardX, scoreCardY - 60, 160, 60);
            context.currentStream.fill();
            
            // Side line depending on score color
            if (score >= 80) {
                context.setNonStrokingColor(34, 197, 94); // Green
            } else if (score >= 60) {
                context.setNonStrokingColor(234, 179, 8); // Yellow
            } else {
                context.setNonStrokingColor(239, 68, 68); // Red
            }
            context.currentStream.addRect(scoreCardX, scoreCardY - 60, 8, 60);
            context.currentStream.fill();
            
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 9);
            context.setNonStrokingColor(100, 116, 139); // Slate 500
            context.currentStream.newLineAtOffset(scoreCardX + 20, scoreCardY - 20);
            context.currentStream.showText("OVERALL ATS SCORE");
            
            context.currentStream.setFont(titleFont, 24);
            if (score >= 80) {
                context.setNonStrokingColor(22, 163, 74);
            } else if (score >= 60) {
                context.setNonStrokingColor(202, 138, 4);
            } else {
                context.setNonStrokingColor(220, 38, 38);
            }
            context.currentStream.newLineAtOffset(0, -25);
            context.currentStream.showText(score + "%");
            context.currentStream.endText();
            
            context.currentY -= 110;
            
            // 3. Feedback Summary
            context.checkNewPage(100);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(30, 41, 59); // Slate 800
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Feedback Summary");
            context.currentStream.endText();
            context.currentY -= 18;
            
            String summary = resume.getAnalysis() != null ? resume.getAnalysis().getFeedbackSummary() : "No summary available.";
            context.currentY = drawWrappedText(context, summary, context.margin, context.currentY, context.printableWidth, bodyFont, 10, 14);
            context.currentY -= 15;
            
            // 4. Key Strengths
            context.checkNewPage(80);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(22, 163, 74); // Green
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Key Strengths");
            context.currentStream.endText();
            context.currentY -= 16;
            
            List<String> strengths = (resume.getAnalysis() != null && resume.getAnalysis().getStrengths() != null) 
                    ? resume.getAnalysis().getStrengths() : List.of("Readable text structure");
            for (String strength : strengths) {
                context.checkNewPage(30);
                context.currentY = drawWrappedText(context, "• " + strength, context.margin + 10, context.currentY, context.printableWidth - 10, bodyFont, 10, 14);
            }
            context.currentY -= 15;
            
            // 5. Areas of Improvement
            context.checkNewPage(80);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(220, 38, 38); // Red
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Areas of Improvement");
            context.currentStream.endText();
            context.currentY -= 16;
            
            List<String> improvements = (resume.getAnalysis() != null && resume.getAnalysis().getImprovements() != null) 
                    ? resume.getAnalysis().getImprovements() : List.of("Add more quantifiable achievements");
            for (String improvement : improvements) {
                context.checkNewPage(30);
                context.currentY = drawWrappedText(context, "• " + improvement, context.margin + 10, context.currentY, context.printableWidth - 10, bodyFont, 10, 14);
            }
            context.currentY -= 15;
            
            // 6. Recommended Skills
            context.checkNewPage(80);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(37, 99, 235); // Blue
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Recommended Skills to Add");
            context.currentStream.endText();
            context.currentY -= 16;
            
            List<String> skills = (resume.getAnalysis() != null && resume.getAnalysis().getRecommendedSkills() != null) 
                    ? resume.getAnalysis().getRecommendedSkills() : new ArrayList<>();
            if (skills.isEmpty()) {
                context.currentY = drawWrappedText(context, "No additional skills are recommended at this time.", context.margin + 10, context.currentY, context.printableWidth - 10, bodyFont, 10, 14);
            } else {
                String skillsStr = String.join(", ", skills);
                context.currentY = drawWrappedText(context, skillsStr, context.margin + 10, context.currentY, context.printableWidth - 10, bodyFont, 10, 14);
            }
            
            // Add a footer line on the current active page
            context.setNonStrokingColor(148, 163, 184); // Slate 400
            context.currentStream.setLineWidth(0.5f);
            context.setStrokingColor(226, 232, 240); // Slate 200
            context.currentStream.moveTo(context.margin, 60);
            context.currentStream.lineTo(context.margin + context.printableWidth, 60);
            context.currentStream.stroke();
            
            context.currentStream.beginText();
            context.currentStream.setFont(bodyFont, 8);
            context.currentStream.newLineAtOffset(context.margin, 45);
            context.currentStream.showText("Report generated automatically by AI Prep Platform. Evaluated using Llama 3.3.");
            context.currentStream.endText();
            
            // Close context streams
            context.close();
            
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            document.save(out);
            return out.toByteArray();
        }
    }

    public byte[] generateJdMatchReport(Resume resume, String jobDescription, Map<String, Object> matchResult) throws IOException {
        log.info("Generating Job Match PDF report for resume: {}", resume.getId());
        
        try (PDDocument document = new PDDocument()) {
            DocumentContext context = new DocumentContext(document);
            
            PDType1Font titleFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA_BOLD);
            PDType1Font headerFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA_BOLD);
            PDType1Font bodyFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA);
            PDType1Font bodyBoldFont = new PDType1Font(Standard14Fonts.FontName.HELVETICA_BOLD);
            
            // 1. Draw Title Header Block
            context.checkNewPage(80);
            context.setNonStrokingColor(30, 41, 59); // Slate 800
            context.currentStream.addRect(context.margin, context.currentY - 60, context.printableWidth, 60);
            context.currentStream.fill();
            
            context.currentStream.beginText();
            context.currentStream.setFont(titleFont, 18);
            context.setNonStrokingColor(255, 255, 255); // White
            context.currentStream.newLineAtOffset(context.margin + 20, context.currentY - 38);
            context.currentStream.showText("RESUME & JOB ALIGNMENT REPORT");
            context.currentStream.endText();
            context.currentY -= 80;
            
            // 2. Draw Candidate Details & Match Score
            context.checkNewPage(120);
            
            // Background card for candidate details
            context.setNonStrokingColor(248, 250, 252); // Slate 50
            context.currentStream.addRect(context.margin, context.currentY - 90, context.printableWidth, 90);
            context.currentStream.fill();
            
            // Border line for card
            context.setNonStrokingColor(99, 102, 241); // Indigo 500
            context.currentStream.addRect(context.margin, context.currentY - 90, 5, 90);
            context.currentStream.fill();
            
            // Draw details text
            context.currentStream.beginText();
            context.currentStream.setFont(bodyBoldFont, 12);
            context.setNonStrokingColor(51, 65, 85); // Slate 700
            context.currentStream.newLineAtOffset(context.margin + 20, context.currentY - 25);
            context.currentStream.showText("Candidate Details");
            
            context.currentStream.setFont(bodyFont, 10);
            context.setNonStrokingColor(71, 85, 105);
            
            String name = resume.getParsedData() != null ? resume.getParsedData().getName() : "Not Found";
            String email = resume.getParsedData() != null ? resume.getParsedData().getEmail() : "Not Found";
            
            context.currentStream.newLineAtOffset(0, -18);
            context.currentStream.showText("Name: " + name);
            context.currentStream.newLineAtOffset(0, -15);
            context.currentStream.showText("Email: " + email);
            context.currentStream.newLineAtOffset(0, -15);
            context.currentStream.showText("Resume File: " + resume.getFileName());
            context.currentStream.endText();
            
            // Match Score Box
            int score = (matchResult.get("matchScore") != null) ? (Integer) matchResult.get("matchScore") : 50;
            
            float scoreCardX = context.margin + 300;
            float scoreCardY = context.currentY - 15;
            
            context.setNonStrokingColor(241, 245, 249); // Slate 100
            context.currentStream.addRect(scoreCardX, scoreCardY - 60, 160, 60);
            context.currentStream.fill();
            
            if (score >= 80) {
                context.setNonStrokingColor(34, 197, 94); // Green
            } else if (score >= 60) {
                context.setNonStrokingColor(234, 179, 8); // Yellow
            } else {
                context.setNonStrokingColor(239, 68, 68); // Red
            }
            context.currentStream.addRect(scoreCardX, scoreCardY - 60, 8, 60);
            context.currentStream.fill();
            
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 9);
            context.setNonStrokingColor(100, 116, 139);
            context.currentStream.newLineAtOffset(scoreCardX + 20, scoreCardY - 20);
            context.currentStream.showText("JOB COMPATIBILITY");
            
            context.currentStream.setFont(titleFont, 24);
            if (score >= 80) {
                context.setNonStrokingColor(22, 163, 74);
            } else if (score >= 60) {
                context.setNonStrokingColor(202, 138, 4);
            } else {
                context.setNonStrokingColor(220, 38, 38);
            }
            context.currentStream.newLineAtOffset(0, -25);
            context.currentStream.showText(score + "%");
            context.currentStream.endText();
            
            context.currentY -= 110;
            
            // 3. Fit Analysis
            context.checkNewPage(100);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(30, 41, 59);
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Fit Analysis");
            context.currentStream.endText();
            context.currentY -= 18;
            
            String fitAnalysis = (String) matchResult.get("fitAnalysis");
            if (fitAnalysis == null) fitAnalysis = "No fit analysis details available.";
            context.currentY = drawWrappedText(context, fitAnalysis, context.margin, context.currentY, context.printableWidth, bodyFont, 10, 14);
            context.currentY -= 15;
            
            // 4. Matched Keywords
            context.checkNewPage(100);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(22, 163, 74); // Green
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Matched Keywords / Skills");
            context.currentStream.endText();
            context.currentY -= 18;
            
            List<String> matched = (List<String>) matchResult.get("matchedKeywords");
            if (matched == null || matched.isEmpty()) {
                context.currentY = drawWrappedText(context, "No direct matching keywords identified.", context.margin, context.currentY, context.printableWidth, bodyFont, 10, 14);
            } else {
                String matchedStr = String.join(", ", matched);
                context.currentY = drawWrappedText(context, matchedStr, context.margin, context.currentY, context.printableWidth, bodyFont, 10, 14);
            }
            context.currentY -= 15;
            
            // 5. Missing Keywords
            context.checkNewPage(100);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(220, 38, 38); // Red
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Missing / Weak Keywords");
            context.currentStream.endText();
            context.currentY -= 18;
            
            List<String> missing = (List<String>) matchResult.get("missingKeywords");
            if (missing == null || missing.isEmpty()) {
                context.currentY = drawWrappedText(context, "No core missing keywords identified.", context.margin, context.currentY, context.printableWidth, bodyFont, 10, 14);
            } else {
                String missingStr = String.join(", ", missing);
                context.currentY = drawWrappedText(context, missingStr, context.margin, context.currentY, context.printableWidth, bodyFont, 10, 14);
            }
            context.currentY -= 15;
            
            // 6. Tailoring Recommendations
            context.checkNewPage(100);
            context.currentStream.beginText();
            context.currentStream.setFont(headerFont, 12);
            context.setNonStrokingColor(37, 99, 235); // Blue
            context.currentStream.newLineAtOffset(context.margin, context.currentY);
            context.currentStream.showText("Tailoring Recommendations");
            context.currentStream.endText();
            context.currentY -= 18;
            
            List<String> recs = (List<String>) matchResult.get("recommendations");
            if (recs == null || recs.isEmpty()) {
                context.currentY = drawWrappedText(context, "No recommendations available.", context.margin, context.currentY, context.printableWidth, bodyFont, 10, 14);
            } else {
                for (String rec : recs) {
                    context.checkNewPage(30);
                    context.currentY = drawWrappedText(context, "• " + rec, context.margin + 10, context.currentY, context.printableWidth - 10, bodyFont, 10, 14);
                }
            }
            
            // Footer
            context.setNonStrokingColor(148, 163, 184); // Slate 400
            context.currentStream.setLineWidth(0.5f);
            context.setStrokingColor(226, 232, 240); // Slate 200
            context.currentStream.moveTo(context.margin, 60);
            context.currentStream.lineTo(context.margin + context.printableWidth, 60);
            context.currentStream.stroke();
            
            context.currentStream.beginText();
            context.currentStream.setFont(bodyFont, 8);
            context.currentStream.newLineAtOffset(context.margin, 45);
            context.currentStream.showText("Report generated automatically by AI Prep Platform. Matches calculated with Groq AI.");
            context.currentStream.endText();
            
            context.close();
            
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            document.save(out);
            return out.toByteArray();
        }
    }
    
    private float drawWrappedText(DocumentContext context, String text, float x, float y, float width, 
                                  PDType1Font font, float fontSize, float leading) throws IOException {
        context.currentStream.setFont(font, fontSize);
        String[] words = text.split("\\s+");
        StringBuilder line = new StringBuilder();
        
        for (String word : words) {
            String testLine = line.length() == 0 ? word : line + " " + word;
            float lineWidth = fontSize * font.getStringWidth(testLine) / 1000f;
            if (lineWidth > width) {
                // Check if page needs to be swapped
                context.checkNewPage(leading + 10);
                if (context.currentY == context.pageHeight - context.margin) {
                    // Reset y since a new page was added
                    y = context.currentY;
                }
                
                context.currentStream.beginText();
                context.currentStream.setFont(font, fontSize);
                context.currentStream.newLineAtOffset(x, y);
                context.currentStream.showText(line.toString());
                context.currentStream.endText();
                
                y -= leading;
                context.currentY = y;
                line = new StringBuilder(word);
            } else {
                line.append(line.length() == 0 ? "" : " ").append(word);
            }
        }
        
        if (line.length() > 0) {
            context.checkNewPage(leading + 10);
            if (context.currentY == context.pageHeight - context.margin) {
                y = context.currentY;
            }
            
            context.currentStream.beginText();
            context.currentStream.setFont(font, fontSize);
            context.currentStream.newLineAtOffset(x, y);
            context.currentStream.showText(line.toString());
            context.currentStream.endText();
            
            y -= leading;
            context.currentY = y;
        }
        
        return y;
    }
}

File: backend/src/services/pdf-report.service.js
const PDFDocument = require('pdfkit');

const generateResumeReport = (resume) => {
  return new Promise((resolve, reject) => {
    try {
      const doc = new PDFDocument({ size: 'A4', margin: 50 });
      const chunks = [];
      
      doc.on('data', (chunk) => chunks.push(chunk));
      doc.on('end', () => resolve(Buffer.concat(chunks)));
      doc.on('error', (err) => reject(err));
      
      // 1. Draw Title Header Block
      doc.rect(50, 50, 495, 60).fill('#1E293B'); // Slate 800
      doc.fillColor('#FFFFFF')
         .font('Helvetica-Bold')
         .fontSize(18)
         .text('RESUME ATS ANALYSIS REPORT', 70, 72);
      
      // 2. Draw Candidate Details & ATS Score Card
      doc.rect(50, 130, 495, 90).fill('#F8FAFC'); // Slate 50
      doc.rect(50, 130, 5, 90).fill('#CBD5E1'); // Slate 300
      
      const parsedData = resume.parsedData || {};
      const name = parsedData.name || 'Not Found';
      const email = parsedData.email || 'Not Found';
      const phone = parsedData.phone || 'Not Found';
      
      doc.fillColor('#334155')
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Candidate Details', 70, 142);
         
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10)
         .text(`Name: ${name}`, 70, 162)
         .text(`Email: ${email}`, 70, 178)
         .text(`Phone: ${phone}`, 70, 194);
         
      // ATS Score Badge
      const analysis = resume.analysis || {};
      const score = analysis.atsScore || 70;
      
      let scoreColor = '#EF4444'; // Red
      if (score >= 80) scoreColor = '#22C55E'; // Green
      else if (score >= 60) scoreColor = '#EAB308'; // Yellow
      
      doc.rect(350, 145, 160, 60).fill('#F1F5F9'); // Slate 100
      doc.rect(350, 145, 8, 60).fill(scoreColor);
      
      doc.fillColor('#64748B')
         .font('Helvetica-Bold')
         .fontSize(9)
         .text('OVERALL ATS SCORE', 370, 155);
         
      doc.fillColor(scoreColor)
         .font('Helvetica-Bold')
         .fontSize(24)
         .text(`${score}%`, 370, 172);
         
      // 3. Feedback Summary
      doc.fillColor('#1E293B')
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Feedback Summary', 50, 245);
         
      const summary = analysis.feedbackSummary || 'No summary available.';
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10)
         .text(summary, 50, 265, { width: 495, lineGap: 4 });
         
      doc.moveDown(1.5);
      
      // 4. Key Strengths
      doc.fillColor('#16A34A') // Green
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Key Strengths');
         
      doc.moveDown(0.5);
      
      const strengths = analysis.strengths || ['Readable text structure'];
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10);
      strengths.forEach(strength => {
        doc.text(`• ${strength}`, { width: 495, lineGap: 3 });
      });
      
      doc.moveDown(1.5);
      
      // 5. Areas of Improvement
      doc.fillColor('#DC2626') // Red
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Areas of Improvement');
         
      doc.moveDown(0.5);
      
      const improvements = analysis.improvements || ['Add more quantifiable achievements'];
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10);
      improvements.forEach(imp => {
        doc.text(`• ${imp}`, { width: 495, lineGap: 3 });
      });
      
      doc.moveDown(1.5);
      
      // 6. Recommended Skills
      doc.fillColor('#2563EB') // Blue
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Recommended Skills to Add');
         
      doc.moveDown(0.5);
      
      const skills = analysis.recommendedSkills || [];
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10);
      if (skills.length === 0) {
        doc.text('No additional skills are recommended at this time.');
      } else {
        doc.text(skills.join(', '), { width: 495, lineGap: 3 });
      }
      
      // Footer
      const finalPageY = 750;
      doc.strokeColor('#E2E8F0')
         .lineWidth(0.5)
         .moveTo(50, finalPageY)
         .lineTo(545, finalPageY)
         .stroke();
         
      doc.fillColor('#94A3B8')
         .font('Helvetica')
         .fontSize(8)
         .text('Report generated automatically by AI Prep Platform. Evaluated using Llama 3.3.', 50, finalPageY + 15);
         
      doc.end();
    } catch (err) {
      reject(err);
    }
  });
};

const generateJdMatchReport = (resume, jobDescription, matchResult) => {
  return new Promise((resolve, reject) => {
    try {
      const doc = new PDFDocument({ size: 'A4', margin: 50 });
      const chunks = [];
      
      doc.on('data', (chunk) => chunks.push(chunk));
      doc.on('end', () => resolve(Buffer.concat(chunks)));
      doc.on('error', (err) => reject(err));
      
      // 1. Draw Title Header Block
      doc.rect(50, 50, 495, 60).fill('#1E293B'); // Slate 800
      doc.fillColor('#FFFFFF')
         .font('Helvetica-Bold')
         .fontSize(18)
         .text('RESUME & JOB ALIGNMENT REPORT', 70, 72);
         
      // 2. Draw Candidate Details & Match Score Card
      doc.rect(50, 130, 495, 90).fill('#F8FAFC'); // Slate 50
      doc.rect(50, 130, 5, 90).fill('#6366F1'); // Indigo 500
      
      const parsedData = resume.parsedData || {};
      const name = parsedData.name || 'Not Found';
      const email = parsedData.email || 'Not Found';
      
      doc.fillColor('#334155')
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Candidate Details', 70, 142);
         
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10)
         .text(`Name: ${name}`, 70, 162)
         .text(`Email: ${email}`, 70, 178)
         .text(`Resume File: ${resume.fileName || 'Not Found'}`, 70, 194);
         
      // Match Score Box
      const score = matchResult.matchScore || 50;
      
      let scoreColor = '#EF4444'; // Red
      if (score >= 80) scoreColor = '#22C55E'; // Green
      else if (score >= 60) scoreColor = '#EAB308'; // Yellow
      
      doc.rect(350, 145, 160, 60).fill('#F1F5F9'); // Slate 100
      doc.rect(350, 145, 8, 60).fill(scoreColor);
      
      doc.fillColor('#64748B')
         .font('Helvetica-Bold')
         .fontSize(9)
         .text('JOB COMPATIBILITY', 370, 155);
         
      doc.fillColor(scoreColor)
         .font('Helvetica-Bold')
         .fontSize(24)
         .text(`${score}%`, 370, 172);
         
      // 3. Fit Analysis
      doc.fillColor('#1E293B')
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Fit Analysis', 50, 245);
         
      const fitAnalysis = matchResult.fitAnalysis || 'No fit analysis details available.';
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10)
         .text(fitAnalysis, 50, 265, { width: 495, lineGap: 4 });
         
      doc.moveDown(1.5);
      
      // 4. Matched Keywords
      doc.fillColor('#16A34A') // Green
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Matched Keywords / Skills');
         
      doc.moveDown(0.5);
      
      const matched = matchResult.matchedKeywords || [];
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10);
      if (matched.length === 0) {
        doc.text('No direct matching keywords identified.');
      } else {
        doc.text(matched.join(', '), { width: 495, lineGap: 3 });
      }
      
      doc.moveDown(1.5);
      
      // 5. Missing Keywords
      doc.fillColor('#DC2626') // Red
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Missing / Weak Keywords');
         
      doc.moveDown(0.5);
      
      const missing = matchResult.missingKeywords || [];
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10);
      if (missing.length === 0) {
        doc.text('No core missing keywords identified.');
      } else {
        doc.text(missing.join(', '), { width: 495, lineGap: 3 });
      }
      
      doc.moveDown(1.5);
      
      // 6. Tailoring Recommendations
      doc.fillColor('#2563EB') // Blue
         .font('Helvetica-Bold')
         .fontSize(12)
         .text('Tailoring Recommendations');
         
      doc.moveDown(0.5);
      
      const recs = matchResult.recommendations || [];
      doc.fillColor('#475569')
         .font('Helvetica')
         .fontSize(10);
      if (recs.length === 0) {
        doc.text('No recommendations available.');
      } else {
        recs.forEach(rec => {
          doc.text(`• ${rec}`, { width: 495, lineGap: 3 });
        });
      }
      
      // Footer
      const finalPageY = 750;
      doc.strokeColor('#E2E8F0')
         .lineWidth(0.5)
         .moveTo(50, finalPageY)
         .lineTo(545, finalPageY)
         .stroke();
         
      doc.fillColor('#94A3B8')
         .font('Helvetica')
         .fontSize(8)
         .text('Report generated automatically by AI Prep Platform. Matches calculated with Groq AI.', 50, finalPageY + 15);
         
      doc.end();
    } catch (err) {
      reject(err);
    }
  });
};

module.exports = {
  generateResumeReport,
  generateJdMatchReport
};
